### PR TITLE
`parse_factor()` gains a `include_na` argument

### LINKED ISSUE #541: parse_factor should always generate explicit NAs
Or it should at least be an option

### PR DESCRIPTION
**Pull request by jimhester**
Which includes `NA` in the factor levels.

Fixes #541

### CODE REVIEW

File: R/collectors.R [hunk 1]
```diff
@@ -200,6 +200,7 @@ guess_parser <- function(x, locale = default_locale()) {
 #'
 #' @param levels Character vector providing set of allowed levels.
 #' @param ordered Is it an ordered factor?
+#' @param include_na Should `NA` be included as a factor level?
```
Comment (hadley) [inline R/collectors.R]: I think this should be, "If `NA` are present, include as an explicit factor to level?".  And I think it should default to `TRUE`

File: tests/testthat/test-parsing-factors.R [hunk 2]
```diff
@@ -22,3 +22,7 @@ test_that("NAs silently passed along", {
   expect_equal(x, factor(c("a", "b", NA)))
 })
 
+test_that("NAs included in levels if desired", {
+  x <- parse_factor(c("a", "b", "NA"), levels = c("a", "b"), include_na = TRUE)
```
Comment (hadley) [inline tests/testthat/test-parsing-factors.R]: Can you also include a test with `levels = NULL`?

Comment (jimhester) [inline tests/testthat/test-parsing-factors.R]: I can once https://github.com/tidyverse/readr/pull/602 is done

File: R/collectors.R (hunk 1 above)
Comment (jimhester) [inline R/collectors.R]: The current implementation includes NA regardless if it occurs in the input, I assume you would want it to only be included if the input contains NA?

Comment (hadley) [inline R/collectors.R]: Right

**Comment by jimhester** [comment]
Closing in favor of https://github.com/tidyverse/readr/pull/617
